public/app/percona/shared/components/hooks/cancelToken.hook.ts

```diff
@@ -0,0 +1,23 @@
+import { useRef, useEffect } from 'react';
+import axios, { CancelTokenSource } from 'axios';
+
+export const useCancelToken = () => {
+  const tokens = useRef<Record<string, CancelTokenSource>>({});
+
+  const generateToken = (sourceName: string) => {
```

Comment:
what about moving `sourceName` to `useCancelToken` to then return the token itself?

it would be used like this:

```
const [cancelToken1] = useCancelToken(MY_TOKEN_1)
const [cancelToken2] = useCancelToken(MY_TOKEN_2)
```

the cleanup would run twice in the example above, but on half the set of tokens.

Reply:
This come with the problem that, using that method, you won't be able to cancel the previous similar request (which is taken care of by the hook now using the current approach), because you have to generate a new token when you cancel the previous one, which you can't accomplish the way you mentioned.

Reply:
Not sure I'm following you. One problem is that a lot is getting hidden here.

why do we need to generate the token every time? if the component gets unmounted the generated token gets already canceled, am I missing something?

Reply:
That's only one use-case (cancel all requests on unmount).
There's another, which you can see in the second screenshot I put there: to cancel eventual previous identical requests. That's why you pass a string constant (`MY_TOKEN_1 `, `MY_TOKEN_2`, ...) to `generateToken()`: it's the way the hook keeps track of current requests. If a request with the same string will be triggered, the hook cancels the previous one - first line of `generateToken` function.